### PR TITLE
Develop to Main

### PR DESCRIPTION
**Pull request by YarinsHyper**

### CODE REVIEW

File: controller/service.go [hunk 1]
```diff
@@ -12,24 +12,30 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+//Service is a type of every method
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// Service is a type of every method
```
Add a space between the`//` and the first letter -for better readability. Do the same with all comments.

File: controller/service.go [hunk 2]
```diff
@@ -12,24 +12,30 @@ import (
...
 	Name           string
 	Date           string
 	PersonalNumber string
 	ID             string `bson:"_id" json:"id,omitempty"`
 }
 
+//NewService is used each time i create a method
 func NewService(url string) *Service {
 	s := &Service{}
+	s.BirthdayCollection = mongoConnect.CNX.Database("birthdayDB").Collection("birthdays")
```
Comment (Shahar-Y) [inline controller/service.go:43]: "birthdays" and "birthdayDB" should be a const

File: controller/service.go [hunk 3]
```diff
@@ -12,24 +12,30 @@ import (
...
+
 	return s
 }
 
-func (bs *BirthdayStruct) asBirthdayObject() *pb.BirthdayObject {
+func (bs *BirthdayModel) asBirthdayObject() *pb.BirthdayObject {
```
Comment (Shahar-Y) [inline controller/service.go:49]: add documentation even if not public

File: controller/service.go [hunk 4]
```diff
@@ -63,31 +69,56 @@ func (s *Service) CreateBirthday(ctx context.Context, req *pb.CreateBirthdayRequ
 	return convertedBirthday, nil
 }
 
+//GetBirthday is getting a birthday object
```
Comment (Shahar-Y) [inline controller/service.go]: `getting a birthday object` from what? by it's ID? private number? - it's not clear for an outside-observer. 
same for the other methods

File: controller/service.go [hunk 5]
```diff
@@ -63,31 +69,56 @@ func (s *Service) CreateBirthday(ctx context.Context, req *pb.CreateBirthdayRequ
...
+	birthday := s.BirthdayCollection.FindOne(ctx, bson.M{"personalNumber": req.PersonalNumber})
+	BirthdayModel := &BirthdayModel{}
 	birthday.Decode(BirthdayModel)
 	convertedBirthday := BirthdayModel.asBirthdayObject()
 
 	return convertedBirthday, nil
 }
 
+//GetAllBirthday is creating all birthday objects
+func (s *Service) GetAllBirthday(ctx context.Context, req *pb.GetAllBirthdayRequest) (*pb.GetAllBirthdayResponse, error) {
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
func (s *Service) GetAllBirthdays(ctx context.Context, req *pb.GetAllBirthdayRequest) (*pb.GetAllBirthdayResponse, error) {
```

File: controller/service.go [hunk 6]
```diff
@@ -63,31 +69,56 @@ func (s *Service) CreateBirthday(ctx context.Context, req *pb.CreateBirthdayRequ
...
+
+// UpdateBirthday is updating a birthday object
 func (s *Service) UpdateBirthday(ctx context.Context, req *pb.UpdateBirthdayRequest) (*pb.BirthdayObject, error) {
-	birthdayCollection := mongoConnect.CNX.Database("birthdayDB").Collection("birthdays")
 
 	opts := options.Update().SetUpsert(true)
-	filter := bson.D{{"personalNumber", req.PersonalNumber}}
-	update := bson.D{{"$set", bson.D{{"name", req.Name}, {"date", req.Date}, {"personalNumber", req.PersonalNumber}}}}
-	result, err := birthdayCollection.UpdateOne(context.TODO(), filter, update, opts)
+	filter := bson.D{{Key: "personalNumber", Value: req.PersonalNumber}}
```
Comment (Shahar-Y) [inline controller/service.go]: All of the strings here should be consts as labels. See the constants in permission-service's `store.go`

File: controller/service.go [hunk 7]
```diff
@@ -96,10 +127,10 @@ func (s *Service) UpdateBirthday(ctx context.Context, req *pb.UpdateBirthdayRequ
 	return convertedBirthday, nil
 }
 
+// DeleteBirthday is deleting a birthday object
 func (s *Service) DeleteBirthday(ctx context.Context, req *pb.DeleteBirthdayRequest) (*pb.DeleteBirthdayResponse, error) {
-	birthdayCollection := mongoConnect.CNX.Database("birthdayDB").Collection("birthdays")
 
-	result, err := birthdayCollection.DeleteOne(ctx, bson.M{"personalNumber": req.PersonalNumber})
+	result, err := s.BirthdayCollection.DeleteOne(ctx, bson.M{"personalNumber": req.PersonalNumber})
```
Comment (Shahar-Y) [inline controller/service.go]: "personalNumber" should be a const

File: mongoDB/mongo.manager.go [hunk 8]
```diff
@@ -9,8 +9,11 @@ import (
...
 )
 
+//CNX is a mongo client
 var CNX = Connection()
 
+//Connection is a function that creates a mongo client and
+//connects it to the db and pings/checks the connection
 func Connection() *mongo.Client {
 	// Set client options
 	clientOptions := options.Client().ApplyURI("mongodb://root:example@0.0.0.0:27017")
```
Comment (Shahar-Y) [inline mongoDB/mongo.manager.go]: Should be an environment variable, and the port should be a different envar.
See `server.go` in Drive's api-gateway and use viper

File: proto/birthday.proto [hunk 9]
```diff
@@ -5,66 +5,72 @@ option go_package = "github.com/yarinBenisty/birthday-service/proto/birthday.pro
...
 
 message CreateBirthdayRequest {
-    //the name of the person
-    string name = 1;
+  // the name of the person
+  string name = 1;
 
-    //the date of birth of the person
-    string date = 2;
+  // the date of birth of the person
```
Comment (Shahar-Y) [inline proto/birthday.proto]: How is the birth-date represented? In milliseconds?
Also, why is it not an int64?

File: proto/birthday.proto [hunk 10]
```diff
@@ -5,66 +5,72 @@ option go_package = "github.com/yarinBenisty/birthday-service/proto/birthday.pro
...
 
-    //the name of the person
-    string name = 2;
+  // the name of the person
+  string name = 2;
 
-    //the date of birth of the person
-    string date = 3;
+  // the date of birth of the person
+  string date = 3;
```
Comment (Shahar-Y) [inline proto/birthday.proto:53]: Why is date not int64?

File: proto/birthday.proto [hunk 11]
```diff
@@ -5,66 +5,72 @@ option go_package = "github.com/yarinBenisty/birthday-service/proto/birthday.pro
...
+  rpc UpdateBirthday(UpdateBirthdayRequest) returns (BirthdayObject) {}
 
+  // DeleteBirthday deletes the selected birthday.
+  rpc DeleteBirthday(DeleteBirthdayRequest) returns (DeleteBirthdayResponse) {}
 }
 
 message CreateBirthdayRequest {
-    //the name of the person
-    string name = 1;
+  // the name of the person
```
Comment (Shahar-Y) [inline proto/birthday.proto]: Start all of the documentations in an upper-case letter
```suggestion
  // The name of the person
```

File: proto/birthday.proto [hunk 12]
```diff
@@ -5,66 +5,72 @@ option go_package = "github.com/yarinBenisty/birthday-service/proto/birthday.pro
...
 
 service BirthdayFunctions {
-    // CreateBirthday creates a new birthday and returns it.
-    rpc CreateBirthday(CreateBirthdayRequest) returns (BirthdayObject) {}
+  // CreateBirthday creates a new birthday and returns it.
+  rpc CreateBirthday(CreateBirthdayRequest) returns (BirthdayObject) {}
 
-    // GetBirthday returns the selected birthday.
-    rpc GetBirthday(GetBirthdayRequest) returns (BirthdayObject) {}
+  // GetBirthday returns the selected birthday.
```
Comment (Shahar-Y) [inline proto/birthday.proto]: How does it find the birthday? (Using the personal number)
- If it is not part of the name (for examle, `GetBirthdayByPersonalNumber`), then you should document it.

File: controller/service.go [hunk 13]
```diff
@@ -63,31 +69,56 @@ func (s *Service) CreateBirthday(ctx context.Context, req *pb.CreateBirthdayRequ
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	var mongoBirthdays []BirthdayModel
+	if err = cursor.All(ctx, &mongoBirthdays); err != nil {
+		log.Fatal(err)
+	}
+	birthdaysArray := make([]*pb.BirthdayObject, 0, len(mongoBirthdays))
+
+	for index := 0; index < len(mongoBirthdays); index++ {
```
Comment (Shahar-Y) [inline controller/service.go:108]: What does this for loop do? Add documentation.

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Good job! 
Fix according to the comments, and reply `don` on each (or comment back if you have a different reply).
Do not resolve comments
:+1:

File: controller/service.go (hunk 1 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 2 above)
Comment (YarinsHyper) [reply controller/service.go:43]: done

File: controller/service.go (hunk 3 above)
Comment (YarinsHyper) [reply controller/service.go:49]: done

File: controller/service.go (hunk 4 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 5 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: proto/birthday.proto (hunk 9 above)
Comment (YarinsHyper) [reply proto/birthday.proto]: done.
the reason is, we agreed on from the start that it'll be easier to get and pass on both ways strings of dates.

File: proto/birthday.proto (hunk 10 above)
Comment (YarinsHyper) [reply proto/birthday.proto:53]: done.
the reason is, we agreed on from the start that it'll be easier to get and pass on both ways strings of dates.

File: proto/birthday.proto (hunk 11 above)
Comment (YarinsHyper) [reply proto/birthday.proto]: done

File: proto/birthday.proto (hunk 12 above)
Comment (YarinsHyper) [reply proto/birthday.proto]: done

File: controller/service.go (hunk 13 above)
Comment (YarinsHyper) [reply controller/service.go:108]: done

File: controller/service.go (hunk 6 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 7 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: mongoDB/mongo.manager.go (hunk 8 above)
Comment (YarinsHyper) [reply mongoDB/mongo.manager.go]: done

File: controller/service.go [hunk 14]
```diff
@@ -12,24 +12,40 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+const (
+	birthdayDatabase        = "birthdayDB"
+	birthdayCollection      = "birthdays"
+	personalNumberParameter = "personalNumber"
+	dateParameter           = "date"
+	nameParameter           = "name"
```
Comment (Shahar-Y) [inline controller/service.go:20]: Better used with first letters as upper-case and with a short description, but accepted.

File: controller/service.go [hunk 15]
```diff
@@ -12,24 +12,40 @@ import (
...
+)
+
+// Service is a type of every method
```
Comment (Shahar-Y) [inline controller/service.go]: Description is not clear

File: controller/service.go [hunk 16]
```diff
@@ -12,24 +12,40 @@ import (
...
-type BirthdayStruct struct {
+// BirthdayModel is a model for coverting an object for type: birthdayObject
+type BirthdayModel struct {
 	Name           string
 	Date           string
 	PersonalNumber string
 	ID             string `bson:"_id" json:"id,omitempty"`
 }
 
+// NewService is used each time i create a method
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// NewService creates a Service and returns it.
```
The description is wrong.  You only call it once in the `server.go` file.

File: controller/service.go [hunk 17]
```diff
@@ -12,24 +12,40 @@ import (
...
 func NewService(url string) *Service {
 	s := &Service{}
+	// s.BirthdayCollection = mongoConnect.CNX.Database(birthdayDatabase).Collection(birthdayCollection)
+	s.BirthdayCollection = mongoConnect.CNX.Database("birthdayDB").Collection("birthdays")
+
 	return s
 }
 
-func (bs *BirthdayStruct) asBirthdayObject() *pb.BirthdayObject {
+// asBirthdayObject is making any object type to: BirthdayObject type
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// asBirthdayObject converts the BirthdayModel type to: BirthdayObject type
```
It doesn't convert any object, but only `BirthdayModel` type

File: controller/service.go [hunk 18]
```diff
@@ -39,55 +55,79 @@ func (bs *BirthdayStruct) asBirthdayObject() *pb.BirthdayObject {
...
+	BirthdayModel := &BirthdayModel{}
 	birthday.Decode(BirthdayModel)
 
 	convertedBirthday := BirthdayModel.asBirthdayObject()
-	fmt.Println(convertedBirthday)
 
 	return convertedBirthday, nil
 }
 
+// GetBirthday is getting a birthday object by personalNumber parameter
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// GetBirthday returns a birthday object by personalNumber
```

File: controller/service.go [hunk 19]
```diff
@@ -39,55 +55,79 @@ func (bs *BirthdayStruct) asBirthdayObject() *pb.BirthdayObject {
...
-	BirthdayModel := &BirthdayStruct{}
+	birthday := s.BirthdayCollection.FindOne(ctx, bson.M{personalNumberParameter: req.PersonalNumber})
+	BirthdayModel := &BirthdayModel{}
 	birthday.Decode(BirthdayModel)
 	convertedBirthday := BirthdayModel.asBirthdayObject()
 
 	return convertedBirthday, nil
 }
 
+// GetAllBirthdays is creating all birthday objects
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// GetAllBirthdays returns all birthday objects in the DB
```
The documentation was wrong

File: controller/service.go [hunk 20]
```diff
@@ -96,10 +136,10 @@ func (s *Service) UpdateBirthday(ctx context.Context, req *pb.UpdateBirthdayRequ
 	return convertedBirthday, nil
 }
 
+// DeleteBirthday is deleting a birthday object
```
Comment (Shahar-Y) [inline controller/service.go]:
```suggestion
// DeleteBirthday deletes a birthday object by it's personalNumber
```

File: .env [hunk 21]
```diff
@@ -0,0 +1 @@
+MONGO_URL="mongodb://root:example@0.0.0.0:27017"
```
Comment (Shahar-Y) [inline .env]: We do not upload the `.env` files to github. Keep the `.env` on your machine, and use viper defaults

File: mongoDB/mongo.manager.go [hunk 22]
```diff
@@ -1,19 +1,31 @@
...
+
+	// loading the dotenv parameters
+	envErr := godotenv.Load()
+	if envErr != nil {
+		log.Fatal("Error loading .env file")
+	}
+
 	// Set client options
-	clientOptions := options.Client().ApplyURI("mongodb://root:example@0.0.0.0:27017")
+	clientOptions := options.Client().ApplyURI(os.Getenv("MONGO_URL"))
```
Comment (Shahar-Y) [inline mongoDB/mongo.manager.go]: Not having a default value to the envars is bad practice, use Viper

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Most of the changes requre your commit after you understand why I wrote them

File: controller/service.go (hunk 14 above)
Comment (YarinsHyper) [reply controller/service.go:20]: ok..
done

File: controller/service.go (hunk 15 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 16 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 17 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 18 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 19 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: controller/service.go (hunk 20 above)
Comment (YarinsHyper) [reply controller/service.go]: done

File: .env (hunk 21 above)
Comment (YarinsHyper) [reply .env]: done

File: mongoDB/mongo.manager.go (hunk 22 above)
Comment (YarinsHyper) [reply mongoDB/mongo.manager.go]: used viper, done

**Review by yoavrozner — APPROVED** [review]
Well done @yarinBenisty

**Review by Shahar-Y — APPROVED** [review]
Haven't tested it myself. I trust Yarin's checks.
Good job, you can proceed :+1:

**Pull request merged**